### PR TITLE
Fix signature length correction

### PR DESCRIPTION
**Pull request by JamesSmartCell**
Discovered an old bug - sometimes there can be a lot of leading zeros in the signature. This should come as no surprise considering how bitcoin works. It doesn't happen often ...

Fixed.

### CODE REVIEW

**Comment by James-Sangalli** [comment]
@JamesSmartCell when is a signature greater than 65 bytes?

**Comment by JamesSmartCell** [comment]
Good question and well spotted.

It's a phenomena of Two's Compliment. If the value of the final two bytes is greater than 0x80 then the BigInteger byte interpreter assumes a Two's Compliment value, and to keep the byte conversion positive it has to add an extra byte to the value to keep it matching. This breaks the 65 byte manipulation in the signing system. So, if the converted value is greater than 65 we know that BI has converted it into a Two's Compliment value and we just convert it back into the byte value if it were unsigned.

You may also wonder - say if the top two bytes were zero, and the next byte is greater than 0x80, it will grow into the final byte and break the signature system? It doesn't matter because the signing system converts the byte value back to BI and knows the value is unsigned, so it comes out the same.

File: app/src/main/java/io/stormbird/wallet/entity/SignaturePair.java [hunk 1]
```diff
@@ -153,11 +153,15 @@ public SignaturePair(String qrMessage, String timeMessage, String contractAddr)
...
+        if (signature.length < 65)
         {
+            int offset = 65 - signature.length;
             byte[] sigCopy = new byte[65];
-            System.arraycopy(signature, 0, sigCopy, 1, 64);
-            sigCopy[0] = 0;
+            System.arraycopy(signature, 0, sigCopy, offset, 65-offset);
+            for (int i = 0; i < offset; i++)
+            {
+                sigCopy[0] = 0;
```
Comment (hboon) [inline app/src/main/java/io/stormbird/wallet/entity/SignaturePair.java]: Maybe should be:

    sigCopy[i] = 0;

?

Comment (JamesSmartCell) [reply app/src/main/java/io/stormbird/wallet/entity/SignaturePair.java]: Yes it should be. Thanks :) and thanks for looking, even if I'm not sure how you got here. Yes the buffer probably does initialise implicitly to zero, but for a signature it's best to be completely sure.

Be careful of making yourself too useful we don't want you to burn yourself out.

**Comment by JamesSmartCell** [comment]
so ... do the lords deem this request now worthy of inclusion?

**Pull request merged**